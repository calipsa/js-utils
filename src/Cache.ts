import Singleton from './Singleton'

export default class Cache<K, V> extends Singleton {
  map = new Map<K, V>()

  add(idx: K, item: V) {
    this.map.set(idx, item)
  }

  get(idx: K) {
    return this.map.get(idx)
  }

  clear() {
    this.map.clear()
  }
}
