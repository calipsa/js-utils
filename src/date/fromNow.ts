import moment from 'moment-timezone'

interface Lookup {
  [key: string]: string,
}

const lookup: Lookup = {
  'a day ago': 'yesterday',
  'in a day': 'tomorrow',
}

/**
 * A better `fromNow` supporting 'yesterday' & 'tomorrow'
 */
export default (...args: Parameters<typeof moment>) => {
  const m = moment(...args)
  if (!m.isValid()) {
    return 'N/A'
  }

  const formatted = m.fromNow()
  return lookup[formatted] || formatted
}
