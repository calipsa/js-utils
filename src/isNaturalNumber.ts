/**
 * Checks if a value is a natural number,
 * i.e. is an integer greater than zero
 */
export default (num: any): num is number =>
  Number.isInteger(num) && num > 0
