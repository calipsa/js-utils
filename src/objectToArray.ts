import isNonNegativeInteger from './isNonNegativeInteger'

type O = Readonly<{
  [k: number]: any,
}>

export default <Obj extends O, V extends Obj[keyof Obj]>(o: Obj) => {
  const indices = Object.keys(o)
    .map(Number)
    .filter(isNonNegativeInteger)

  const minKey = Math.min(...indices)
  const maxKey = Math.max(...indices)
  const arr: V[] = []

  for (let i = minKey; i <= maxKey; ++i) {
    arr[i] = o[i]
  }

  return arr
}
