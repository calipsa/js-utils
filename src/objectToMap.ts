/**
 * Converts an object to a `Map`
 */
// eslint-disable-next-line @typescript-eslint/ban-types
export default <T extends {}, K extends keyof T>(o: T) =>
  new Map<string, T[K]>(Object.entries(o))
