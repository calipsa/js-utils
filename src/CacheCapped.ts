import Cache from './Cache'

export default class CacheCapped<K, V> extends Cache<K, V> {
  private capacity: number

  constructor(capacity: number) {
    super()
    this.capacity = capacity
  }

  rm(item: K) {
    this.map.delete(item)
  }

  add(idx: K, item: V) {
    if (this.map.size === this.capacity && !this.map.has(idx)) {
      const oldest = this.map.keys().next().value
      this.rm(oldest)
    }
    this.map.set(idx, item)
  }
}
