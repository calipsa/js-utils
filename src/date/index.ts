import fromNow from './fromNow'

export {
  // eslint-disable-next-line import/prefer-default-export
  fromNow,
}
