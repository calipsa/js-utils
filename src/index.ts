import Cache from './Cache'
import CacheCapped from './CacheCapped'
import delay from './delay'
import hasDuplicates from './hasDuplicates'
import isNonNegativeInteger from './isNonNegativeInteger'
import objectToArray from './objectToArray'
import objectToFunction from './objectToFunction'
import objectToFunctionWithBind from './objectToFunctionWithBind'
import Singleton from './Singleton'
import sliceRotate from './sliceRotate'
import toNearestByStep from './toNearestByStep'

export {
  Cache,
  CacheCapped,
  delay,
  hasDuplicates,
  isNonNegativeInteger,
  objectToArray,
  objectToFunction,
  objectToFunctionWithBind,
  Singleton,
  sliceRotate,
  toNearestByStep,
}
