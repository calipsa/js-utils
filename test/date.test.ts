import moment from 'moment-timezone'

import { fromNow } from '../src/date'

function getInputs() {
  const m = moment()
  const date = m.toDate()

  return {
    str: date.toISOString(),
    date,
    rubbish: 'rubbish shit',
    yesterday: m.clone().subtract(1, 'day').toDate(),
    twoDaysAgo: m.clone().subtract(2, 'day').toDate(),
    tomorrow: m.clone().add(1, 'day').toDate(),
    inTwoDays: m.clone().add(2, 'day').toDate(),
  }
}

describe('Date utils', () => {
  for (const [name, input] of Object.entries(getInputs())) {
    it(`Should fromNow ${name} correctly`, () => {
      const result = fromNow(input)
      expect(result).toMatchSnapshot()
    })
  }

  it('Should treat string, Date & moment instance equally', () => {
    const m = moment()
    m.add(2, 'months')
    const date = m.toDate()
    const str = date.toISOString()

    const res1 = fromNow(str)
    const res2 = fromNow(date)
    const res3 = fromNow(m)
    expect(res2).toBe(res1)
    expect(res3).toBe(res1)
  })
})
