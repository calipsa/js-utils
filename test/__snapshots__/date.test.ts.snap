// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Date utils Should fromNow date correctly 1`] = `"a few seconds ago"`;

exports[`Date utils Should fromNow inTwoDays correctly 1`] = `"in 2 days"`;

exports[`Date utils Should fromNow rubbish correctly 1`] = `"N/A"`;

exports[`Date utils Should fromNow str correctly 1`] = `"a few seconds ago"`;

exports[`Date utils Should fromNow tomorrow correctly 1`] = `"tomorrow"`;

exports[`Date utils Should fromNow twoDaysAgo correctly 1`] = `"2 days ago"`;

exports[`Date utils Should fromNow yesterday correctly 1`] = `"yesterday"`;
